src/many-modules/src/_5_data/types.rs

```diff
@@ -1,83 +1,566 @@
+        self.values.get(&index).map(|v| {
+            let value = v.read()?;
+            Ok(value.clone())
+        })
+    }
+
+    pub fn get_mut(&mut self, index: DataIndex) -> Option<Arc<RwLock<DataValue>>> {
+        self.values.get(&index).map(Clone::clone)
+    }
+
+    pub fn delete(&mut self, index: DataIndex) {
+        self.values.remove(&index);
+    }
 }
 
-impl TryFrom<DataValue> for BigInt {
-    type Error = String;
+impl Encode<()> for DataSet {
+    fn encode<W: encode::Write>(
+        &self,
+        e: &mut Encoder<W>,
+        _ctx: &mut (),
+    ) -> Result<(), encode::Error<W::Error>> {
+        e.map(self.values.len() as u64)?;
 
-    fn try_from(value: DataValue) -> Result<Self, Self::Error> {
-        match value {
-            DataValue::Counter(c) => Ok(c.into()),
-            DataValue::Gauge(g) => g.try_into(),
+        for (k, v) in &self.values {
+            let v = v.read().unwrap_or_else(|x| x.into_inner()).clone();
+            e.encode(k)?.encode(v)?;
         }
+        Ok(())
     }
 }
 
-pub type DataValueTypeCounter = u64;
+impl Encode<BTreeSet<DataIndex>> for DataSet {
+    fn encode<W: encode::Write>(
+        &self,
+        e: &mut Encoder<W>,
+        ctx: &mut BTreeSet<DataIndex>,
+    ) -> Result<(), encode::Error<W::Error>> {
+        e.map(ctx.len() as u64)?;
 
-#[derive(Clone, Decode, Encode, Debug)]
-pub enum DataValueTypeGauge {
-    #[n(0)]
-    Int(#[n(0)] i64),
-    #[n(1)]
-    Float(#[n(0)] f64),
-    #[n(2)]
-    BigInt(#[cbor(n(0), decode_with = "decode_bigint", encode_with = "encode_bigint")] BigInt),
+        let it = self.values.iter().filter(|(k, _)| ctx.contains(*k));
+        for (k, v) in it {
+            let v = v.read().unwrap_or_else(|x| x.into_inner()).clone();
+            e.encode(k)?.encode(v)?;
+        }
+
+        Ok(())
+    }
 }
 
-impl TryFrom<DataValueTypeGauge> for BigInt {
-    type Error = String;
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DataValue {
+    Counter(u64),
+    GaugeInt(BigInt),
+    Unknown(CborAny),
+}
+
+impl DataValue {
+    pub fn create(ty: &DataType) -> Self {
+        match ty {
+            DataType::Counter => Self::Counter(0),
+            DataType::GaugeInt => Self::GaugeInt(BigInt::default()),
+        }
+    }
+
+    pub fn as_counter(&self) -> Option<u64> {
+        match self {
+            Self::Counter(v) => Some(*v),
+            _ => None,
+        }
+    }
 
-    fn try_from(value: DataValueTypeGauge) -> Result<Self, Self::Error> {
-        match value {
-            DataValueTypeGauge::Int(i) => Ok(i.into()),
-            DataValueTypeGauge::BigInt(b) => Ok(b),
-            DataValueTypeGauge::Float(_) => {
-                Err("Floats can't be converted to BigInt without loss".into())
+    pub fn as_gauge_int(&self) -> Option<&BigInt> {
+        match self {
+            Self::GaugeInt(v) => Some(v),
+            _ => None,
+        }
+    }
+
+    pub fn convert(&self, ty: DataType) -> Result<Self, ManyError> {
```

Comment:
`convert_unknown`? or `try_into()`?

Reply:
Removed the function. Now it's not needed.